File: core/src/test/java/org/opencds/cqf/cql/ls/core/utility/ConvertersTest.java
package org.opencds.cqf.cql.ls.core.utility;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Test;

public class ConvertersTest {

    @Test
    void should_returnString_when_inputStreamExists() {
        var expected = "The quick brown fox jumps over the lazy dog";
        try {
            var actual =
                    Converters.inputStreamToString(new ByteArrayInputStream(expected.getBytes(StandardCharsets.UTF_8)));
            assertEquals(actual, expected);
        } catch (IOException e) {
            fail("Unexpected exception thrown. {}", e);
        }
    }

    @Test
    void should_returnStringWithLineBreaks_when_inputStreamHasLineBreaksExists() {
        var expected = "the first day in spring –\n" + "a wind from the ocean\n" + "but no ocean in sight";
        try {
            var actual =
                    Converters.inputStreamToString(new ByteArrayInputStream(expected.getBytes(StandardCharsets.UTF_8)));
            assertEquals(actual, expected);
        } catch (IOException e) {
            fail("Unexpected exception thrown. {}", e);
        }
    }

    @Test
    void should_throwIOException_when_inputStreamToStringHasAnIOError() throws IOException {
        InputStream inputStream = mock(InputStream.class);
        when(inputStream.read()).thenThrow(new IOException("Simulated failure"));
        assertThrows(IOException.class, () -> Converters.inputStreamToString(inputStream));
    }

    @Test
    void should_returnSource_when_stringExists() {
        var expected = Converters.stringToSource("The quick brown fox jumps over the lazy dog");
        assertNotNull(expected);
    }

    @Test
    void should_returnSource_when_inputStreamExists() {
        try {
            var expected = Converters.inputStreamToSource(new ByteArrayInputStream(
                    "The quick brown fox jumps over the lazy dog".getBytes(StandardCharsets.UTF_8)));
            assertNotNull(expected);
        } catch (IOException e) {
            fail("Unexpected exception thrown. {}", e);
        }
    }

    @Test
    void should_throwIOException_when_inputStreamToSourceCalledWithNull() throws IOException {
        InputStream inputStream = mock(InputStream.class);
        when(inputStream.read()).thenThrow(new IOException("Simulated failure"));
        assertThrows(IOException.class, () -> Converters.inputStreamToSource(inputStream));
    }
}

File: ls/server/src/test/kotlin/org/opencds/cqf/cql/ls/core/utility/ConvertersTest.kt
package org.opencds.cqf.cql.ls.core.utility

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

class ConvertersTest {
    @Test
    fun should_returnInstance_when_creatingConverter() {
        // Converters is a Kotlin object (singleton); verify it is accessible
        assertNotNull(Converters)
    }

    @Test
    fun should_returnString_when_inputStreamExists() {
        val expected = "The quick brown fox jumps over the lazy dog"
        val actual = Converters.inputStreamToString(ByteArrayInputStream(expected.toByteArray(StandardCharsets.UTF_8)))
        assertEquals(expected, actual)
    }

    @Test
    fun should_returnStringWithLineBreaks_when_inputStreamHasLineBreaksExists() {
        val expected = "the first day in spring –\na wind from the ocean\nbut no ocean in sight"
        val actual = Converters.inputStreamToString(ByteArrayInputStream(expected.toByteArray(StandardCharsets.UTF_8)))
        assertEquals(expected, actual)
    }

    @Test
    fun should_throwIOException_when_inputStreamToStringHasAnIOError() {
        val inputStream =
            object : InputStream() {
                override fun read(): Int = throw IOException("Simulated failure")
            }
        assertThrows(IOException::class.java) { Converters.inputStreamToString(inputStream) }
    }

    @Test
    fun should_returnSource_when_stringExists() {
        val expected = Converters.stringToSource("The quick brown fox jumps over the lazy dog")
        assertNotNull(expected)
    }

    @Test
    fun should_returnSource_when_inputStreamExists() {
        val expected =
            Converters.inputStreamToSource(
                ByteArrayInputStream("The quick brown fox jumps over the lazy dog".toByteArray(StandardCharsets.UTF_8)),
            )
        assertNotNull(expected)
    }

    @Test
    fun should_throwIOException_when_inputStreamToSourceCalledWithNull() {
        val inputStream =
            object : InputStream() {
                override fun read(): Int = throw IOException("Simulated failure")
            }
        assertThrows(IOException::class.java) { Converters.inputStreamToSource(inputStream) }
    }
}
